enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextUpdateApplier.java

```diff
@@ -0,0 +1,365 @@
+            return Pair.of( index, completedLatch );
+        };
+
+        enqueueUpdate( population );
+        return completedLatch;
+    }
+
+    private void enqueueUpdate( FulltextIndexUpdate update ) throws IOException
+    {
+        try
+        {
+            workQueue.put( update );
+        }
+        catch ( InterruptedException e )
+        {
+            throw new IOException( "Fulltext index update failed.", e );
+        }
+    }
+
+    void start()
+    {
+        if ( workerThread != null )
+        {
+            throw new IllegalStateException( workerThread.getName() + " already started." );
+        }
+        workerThread = new ApplierThread( workQueue, log, availabilityGuard );
+        workerThread.start();
+    }
+
+    void stop()
+    {
+        boolean enqueued;
+        do
+        {
+            enqueued = workQueue.offer( STOP_SIGNAL );
+        }
+        while ( !enqueued );
+
+        try
+        {
+            workerThread.join();
+            workerThread = null;
+        }
+        catch ( InterruptedException e )
+        {
+            log.error( "Interrupted before " + workerThread.getName() + " could shut down.", e );
+        }
+    }
+
+    private interface FulltextIndexUpdate
+    {
+        Pair<WritableFulltext,BinaryLatch> applyUpdateAndReturnIndex() throws IOException;
+    }
+
+    private static class ApplierThread extends Thread
+    {
+        private LinkedBlockingQueue<FulltextIndexUpdate> workQueue;
+        private final Log log;
+        private final AvailabilityGuard availabilityGuard;
+
+        ApplierThread( LinkedBlockingQueue<FulltextIndexUpdate> workQueue, Log log,
+                       AvailabilityGuard availabilityGuard )
+        {
+            super( "Fulltext Index Add-On Applier Thread" );
+            this.workQueue = workQueue;
+            this.log = log;
+            this.availabilityGuard = availabilityGuard;
+            setDaemon( true );
+        }
+
+        @Override
+        public void run()
+        {
+            waitForDatabaseToBeAvailable();
+            Set<WritableFulltext> refreshableSet = Collections.newSetFromMap( new IdentityHashMap<>() );
+            List<BinaryLatch> latches = new ArrayList<>();
+
+            FulltextIndexUpdate update;
+            while ( (update = getNextUpdate()) != STOP_SIGNAL )
+            {
+                update = drainQueueAndApplyUpdates( update, refreshableSet, latches );
+                refreshAndClearIndexes( refreshableSet );
+                releaseAndClearLatches( latches );
+
+                if ( update == STOP_SIGNAL )
+                {
+                    return;
+                }
+            }
+        }
+
+        private void waitForDatabaseToBeAvailable()
+        {
+            boolean isAvailable;
+            do
+            {
+                isAvailable = availabilityGuard.isAvailable( 100 );
+            }
+            while ( !isAvailable && !availabilityGuard.isShutdown() );
```

Comment:
not sure how useful `!availabilityGuard.isShutdown()` here since available it will be only in case if it's not shut down already and in the worst case, it can be shut down a moment after you check that shutdown flag anyway